pkg/deviceplugins/vgpu_device_manager.go

```diff
@@ -0,0 +1,499 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 SUSE, LLC.
+ *
+ */
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"kubevirt.io/kubevirt/pkg/util"
+	pluginapi "kubevirt.io/kubevirt/pkg/virt-handler/device-manager/deviceplugin/v1beta1"
+
+	"github.com/harvester/pcidevices/pkg/apis/devices.harvesterhci.io/v1beta1"
+)
+
+const (
+	vgpuPrefix = "MDEV_PCI_RESOURCE"
+)
+
+type VGPUDevicePlugin struct {
+	devs         []*pluginapi.Device
+	server       *grpc.Server
+	socketPath   string
+	stop         <-chan struct{}
+	health       chan deviceHealth
+	devicePath   string
+	resourceName string
+	done         chan struct{}
+	deviceRoot   string
+	initialized  bool
+	lock         *sync.Mutex
+	deregistered chan struct{}
+	starter      *DeviceStarter
+	ctx          context.Context
+}
+
+// Not adding more data to the struct, it's big enough already
+func (dp *VGPUDevicePlugin) GetCount() int {
+	var count int
+	for _, dev := range dp.devs {
+		if dev.Health == pluginapi.Healthy {
+			count++
+		}
+	}
+	logrus.Debugf("found device count %d for plugin %s", count, dp.resourceName)
+	return count
+}
+
+func NewVGPUDevicePlugin(ctx context.Context, vGPUList []string, resourceName string) *VGPUDevicePlugin {
+	serverSock := SocketPath(strings.Replace(resourceName, "/", "-", -1))
+
+	initHandler()
+
+	devs := constructVGPUDPIdevices(vGPUList)
+	dpi := &VGPUDevicePlugin{
+		devs:         devs,
+		socketPath:   serverSock,
+		resourceName: resourceName,
+		devicePath:   v1beta1.MdevRoot,
+		deviceRoot:   util.HostRootMount,
+		health:       make(chan deviceHealth),
+		initialized:  false,
+		lock:         &sync.Mutex{},
+		starter: &DeviceStarter{
+			started:  false,
+			stopChan: make(chan struct{}),
+			backoff:  defaultBackoffTime,
+		},
+		ctx: ctx,
+	}
+	return dpi
+}
+
+func constructVGPUDPIdevices(vGPUList []string) (devs []*pluginapi.Device) {
+	for _, v := range vGPUList {
+		dpiDev := &pluginapi.Device{
+			ID:     v,
+			Health: pluginapi.Unhealthy,
+		}
+		devs = append(devs, dpiDev)
+	}
+	return
+}
+
+// Set Started is used after a call to Start. It's purpose is to set the private starter properly
+func (dp *VGPUDevicePlugin) SetStarted(stop chan struct{}) {
+	c := dp.starter
+	c.stopChan = stop
+	c.started = true
+	logrus.Infof("Started DevicePlugin: %s", dp.resourceName)
+}
+
+func (dp *VGPUDevicePlugin) Started() bool {
+	return dp.starter.started
+}
```

Comment:
Should this require a lock as [gets initialized](https://github.com/ibrokethecloud/pcidevices/blob/55de4d75a7cd9fac885bbad31306ff83e0808bd7/pkg/deviceplugins/vgpu_device_manager.go#L425-L429)?

Reply:
the lock is used in vgpu_controller.go at the time of createOrUpdateDevicePlugin: https://github.com/ibrokethecloud/pcidevices/blob/vgpu/pkg/controller/gpudevice/vgpu_controller.go#L241-L243